viking/src/tools/check.rs

```diff
@@ -144,34 +154,44 @@ fn check_all(
                 decomp_symtab,
                 function,
             )?;
-            if !ok {
+            if matches!(ok, CheckResult::MismatchError) {
                 failed.store(true, std::sync::atomic::Ordering::Relaxed);
             }
+            else if matches!(ok, CheckResult::MatchWarn) {
+                matching_functions.lock().unwrap().push(function.addr);
+            }
 
             Ok(())
         })
     })?;
 
+    let functions_to_update = matching_functions.lock().unwrap();
+    if update_matching {
+        let mut new_functions = functions.iter().cloned().collect_vec();
+        new_functions
+            .iter_mut()
+            .filter(|info| functions_to_update.contains(&info.addr))
```

Comment:
This leads to O(n^2) performance degradation if most functions need to be updated (with n the number of functions). Use a hash table for functions_to_update.